librz/core/serialize_core_seek.c

```diff
@@ -0,0 +1,213 @@
+// SPDX-FileCopyrightText: 2023 Quentin Minster <quentin@minster.io>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_util/rz_serialize.h>
+#include <rz_core.h>
+
+/**
+ * \brief Serialize seek history state and save to a sdb
+ *
+ * \param db sdb to save the state
+ * \param core RzCore instance to save from
+ */
+RZ_API void rz_serialize_core_seek_save(RZ_NONNULL Sdb *db, RZ_NONNULL RzCore *core) {
+	rz_return_if_fail(db && core);
+
+	RzList *list = rz_core_seek_list(core);
+	if (!list) {
+		return;
+	}
+
+	RzListIter *iter;
+	RzCoreSeekItem *undo;
+	rz_list_foreach (list, iter, undo) {
+		PJ *j = pj_new();
+		if (!j) {
+			goto err;
+		}
+		pj_o(j);
+		pj_kn(j, "offset", undo->offset);
+		pj_kn(j, "cursor", undo->cursor);
+		pj_kb(j, "current", undo->is_current);
+		pj_end(j);
+
+		char key[12];
+		sdb_set(db, rz_strf(key, "%" PFMT32d, undo->idx), pj_string(j), 0);
+		pj_free(j);
+	}
+
+err:
+	rz_list_free(list);
+}
+
+enum {
+	SEEK_FIELD_OFFSET,
+	SEEK_FIELD_CURSOR,
+	SEEK_FIELD_CURRENT
+};
+
+/**
+ * \brief Create and initialize a JSON parser for seek history items
+ * \return a RzKeyParser* for seek history items
+ **/
+static RzKeyParser *seek_parser_new(void) {
+	RzKeyParser *parser = rz_key_parser_new();
+	if (!parser) {
+		return NULL;
+	}
+
+	rz_key_parser_add(parser, "offset", SEEK_FIELD_OFFSET);
+	rz_key_parser_add(parser, "cursor", SEEK_FIELD_CURSOR);
+	rz_key_parser_add(parser, "current", SEEK_FIELD_CURRENT);
+
+	return parser;
+}
+
+typedef struct {
+	RzCore *core;
+	RzKeyParser *parser;
+	char *current_key;
+	RzVector /*<RzCoreSeekItem>*/ *vec;
+} SeekLoadCtx;
+
+/**
+ * \brief Load a single seek history item
+ * \param ctx context for loading the item
+ * \param k sdb item key
+ * \param v sdb item value (expected to be JSON)
+ **/
+static bool seek_load_item(SeekLoadCtx *ctx, const char *k, const char *v) {
+	bool ret = false;
+	char *json_str = strdup(v);
+	if (!json_str) {
+		return true;
+	}
+	RzJson *json = rz_json_parse(json_str);
+	if (!json || json->type != RZ_JSON_OBJECT) {
+		goto out_free_str;
+	}
+	RzCoreSeekItem seek_item = { 0 };
+
+	RZ_KEY_PARSER_JSON(ctx->parser, json, child, {
+		case SEEK_FIELD_OFFSET:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			seek_item.offset = child->num.u_value;
+			break;
+		case SEEK_FIELD_CURSOR:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			seek_item.cursor = child->num.s_value;
```

Comment:
why `s_value` instead of `u_value` here?

Reply:
Because `cursor` in `RzCoreSeekItem` is an `int`, so I assumed it may be negative sometimes?